File: Python/programmers/etc/level4/지형_이동.py
from heapq import *

dx, dy = [-1,1,0,0], [0,0,-1,1]

def solution(land, height):
    N = len(land)
    visited = [[False]*N for _ in range(N)]
    q = [(0,0,0)]
    answer = 0

    while q:
        v, x, y = heappop(q)
        if visited[x][y]: continue
        visited[x][y] = True
        answer += v
        for d in range(4):
            nx, ny = x+dx[d], y+dy[d]
            if (0 <= nx < N and 0 <= ny < N and not visited[nx][ny]):
                if abs(land[nx][ny]-land[x][y]) > height:
                    heappush(q, (abs(land[nx][ny]-land[x][y]), nx, ny))
                else:
                    heappush(q, (0, nx, ny))

    return answer


land, height = [[1, 4, 8, 10], [5, 5, 5, 5], [10, 10, 10, 10], [10, 10, 10, 20]], 3
print(solution(land, height)) # 15

land, height = [[10, 11, 10, 11], [2, 21, 20, 10], [1, 20, 21, 11], [2, 1, 2, 1]], 1
print(solution(land, height)) # 18

File: Java/programmers/etc/level4/지형이동.java
package CodeTest.Java.programmers.etc.level4;

import java.util.*;

class Land implements Comparable<Land> {
    int x, y, cost;
    
    Land(int x, int y, int cost) {
        this.x = x;
        this.y = y;
        this.cost = cost;
    }

    @Override
    public int compareTo(Land o) {
        return this.cost-o.cost;
    }
}

public class 지형이동 {
    private static int[] dx = {-1,1,0,0};
    private static int[] dy = {0,0,-1,1};

    public int solution(int[][] land, int height) {
        int N = land.length;
        boolean[][] visited = new boolean[N][N];
        PriorityQueue<Land> q = new PriorityQueue<>();
        q.add(new Land(0, 0, 0));
        
        int answer = 0;

        while (!q.isEmpty()) {
            Land cur = q.poll();
            int x = cur.x, y = cur.y, cost = cur.cost;
            
            if (visited[x][y]) continue;
            visited[x][y] = true;
            answer += cost;

            for (int d=0; d<4; d++) {
                int nx = x+dx[d], ny = y+dy[d];
                if (0 <= nx && nx < N && 0 <= ny && ny < N && !visited[nx][ny]) {
                    if (Math.abs(land[nx][ny]-land[x][y]) > height) {
                        q.add(new Land(nx, ny, Math.abs(land[nx][ny]-land[x][y])));
                    } else {
                        q.add(new Land(nx, ny, 0));
                    }
                }
            }
        }

        return answer;
    }   
}
